src/Utilities/DoubleDicts.jl

```diff
@@ -299,6 +299,55 @@ function Base.keys(d::AbstractDoubleDictInner{F,S}) where {F,S}
     return MOI.ConstraintIndex{F,S}.(keys(d.dict))
 end
 
+"""
+    outer_keys(d::AbstractDoubleDict)
+
+Return an iterator over the outer keys of the `AbstractDoubleDict` `d`.
+Each outer key is a `Tuple{DataType,DataType}` so that a double loop can
+be easily used
+```julia
+for (F, S) in DoubleDicts.outer_keys(dict)
+    for (k, v) in dict[F, S]
+        # ...
+    end
+end
+For performance, it is recommended that the inner loop lies in a separate
+function to gurantee type-stability.
+Some outer keys `(F, S)` might lead to an empty `dict[F, S]`.
+If you only want nonempty `dict[F, S]` use [`outer_keys_vector`](@ref).
+```
+"""
+outer_keys(d::AbstractDoubleDict) = keys(d.dict)
+
+"""
+    outer_keys_vector(d::AbstractDoubleDict; only_nonempty::Bool = true)
+
+Return a vector of outer keys of the `AbstractDoubleDict` `d`.
+If `only_nonempty` is set to true only outer keys that have a nonempty
+set of inner keys will be returned.
+Each outer key is a `Tuple{DataType,DataType}` so that a double loop can
+be easily used
+```julia
+for (F, S) in DoubleDicts.outer_keys_vector(dict)
+    for (k, v) in dict[F, S]
+        # ...
+    end
+end
+For performance, it is recommended that the inner loop lies in a separate
+function to gurantee type-stability.
+If you only want an iterator rather than a vector use [`outer_keys`](@ref).
+```
+"""
+function outer_keys_vector(d::AbstractDoubleDict; only_nonempty::Bool = true)
+    out = Tuple{DataType,DataType}[]
```

Comment:
```suggestion
    out = Tuple{Type,Type}[]
```

We should never use `DataType`, because it has a few quirks
```julia
julia> Union{Int,Float64} isa DataType
false

julia> Union{Int,Float64} isa Type
true
```

Reply:
interesting

Reply:
More specifically, `DataType <: Type`, and `DataType` is the "type" of every concrete type. But using `DataType` in dispatch doesn't work for `Union`s or `UnionAll`.

Basically, there's the type system, which at a high level you can just think of as `Type` or `Type{T}`. But because types are also objects in the language, they themselves need a type:
```julia
julia> isa.(Int, (DataType, Union, UnionAll))
(true, false, false)

julia> isa.(Union{Int,Float64}, (DataType, Union, UnionAll))
(false, true, false)

julia> isa.((Vector{T} where {T}), (DataType, Union, UnionAll))
(false, false, true)
```
So restricting the vector to contain `DataType` doesn't let it hold every possible type in Julia.

... and just for good measure, there's this:
```julia
julia> isa.(Union{}, (DataType, Union, UnionAll))
(false, false, false)

julia> typeof(Union{})
Core.TypeofBottom
```